### ISSUE TITLE
[Bug]: Annotation misplaced when rasterizing and saving as PDF

### ISSUE THREAD

**Issue by jnboehm** [matplotlib/matplotlib#23999]
### Bug summary

Annotating a plot results in the annotation being displayed at the wrong coordinate.  This happen when the following circumstances are met:

- the plot data is being rasterized with `rasterized=True`
- the axis are off with `ax.set_axis_off()`
- the figure is saved as a pdf (otherwise the rasterize argument has no effect)

### Code for reproduction

```python
import matplotlib as mpl
import matplotlib.pyplot as plt
import numpy as np

def test(rasterize, save_ext="pdf"):

    fig, ax = plt.subplots()
    ax.set_axis_off()

    rng = np.random.default_rng(1)
    im = rng.integers(0, 256, size=(10, 10), dtype="uint8")

    loc = [0.05, 0.05]
    ax.plot([loc[0]], [loc[1]], marker="x", c="black", rasterized=rasterize)
    imbox = mpl.offsetbox.OffsetImage(im)
    abox = mpl.offsetbox.AnnotationBbox(
        imbox,
        loc,
    )
    ax.add_artist(abox)

    title = "not " if not rasterize else ""
    title += "rasterized"
    ax.set_title(title)

    name = f"test-{rasterize}.{save_ext}"
    print(name, title)
    fig.savefig(name)


def main():
    test(True, "pdf")


if __name__ == "__main__":
    main()
```

### Actual outcome

The black x marker is visible in the figure, where the annotation should be displayed.

![image](https://user-images.githubusercontent.com/12418563/192136017-4576986b-0159-4d7c-b8c4-980d9e668dc0.png)

### Expected outcome

The output should be the same, whether the figure is rasterized or not. The annotation is displayed at the same coordinate as the marker.  The frame around the OffsetImage should be centered.

![image](https://user-images.githubusercontent.com/12418563/192136051-85dcf268-8204-44ba-8560-ee608d58c49d.png)

### Additional information

Also tested on another Linux installation with python `3.9.13`, matplotlib `3.6.0` (installed with conda) and the problem occurs there as well.

### Operating system

Arch

### Matplotlib Version

3.6.0 (and 3.5.2)

### Matplotlib Backend

QtAgg

### Python version

3.10.6

### Jupyter version

Not used

### Installation

Linux package manager

**Comment by jklymak** [comment]
This is the object rasterizing itself in the pdf backend's dpi instead of the figure dpi.  If you do `fig, ax = plt.subplots(dpi=72)` the output looks as expected.  This is definitely a bug, though how to get this artist to respect the desired dpi for positioning is probably going to be hard.

**Comment by jnboehm** [comment]
Thank you for the response. Is there a way to set the dpi of the pdf backend? I was not aware that it had a (potentially) different dpi value than the figure dpi.  

What confuses me a bit is the fact that it works when the axis are being drawn.  But it doesn't matter when I call `ax.set_axis_off()` (before of after the call to the rasterized function), so I am assuming that the rasterization only takes place when the figure is being saved. But why would the axis artist force the scatter object to use the actually specified figure dpi?

Another tidbit of info is that this also occurs within the svg backend.  So it might not be specific to just the pdf backend.

**Comment by jklymak** [comment]
This occurs regardless of the axis being on or off.  And indeed you can have whatever dpi you want for the rasterization in the pdf.  The issue here is that the offset is being calculated in the specified dpi but then the offset is interpreted at draw time in the the fake 72 dpi assumed by pdf and svg.

**Comment by jnboehm** [comment]
If the axis is left as is (remove the call to `ax.set_axis_off()`, then I get the resulting image that I would expect (hence my initial confusion):

![t](https://user-images.githubusercontent.com/12418563/192552785-5ae1fe43-f9ef-4b1f-882d-60972814d3b1.png)

So to me it seems like something is not being properly set when the axis are not being drawn, or am I confusing two things here?

**Comment by jklymak** [comment]
Apologies, you are correct that it doesn't happen with the axes on.  That is very strange.

**Comment by leejjoon** [comment]
I think the issue is the timing of when `stop_rasterization` is called.

Assumeing 
* a, b are artists with rasterization allowed (artists decorated by allow_rasterization) and c is not.
* only a is rasterized
* the drawing order is a-c-b
 
'start_rasterization' is called when a is drawn, but the `stop_rasterization` is called when b is drawn. Thus c is affected by the dpi change. 

One workaround would be to add a rasterization-allowed, do-nothing artist in between.
```python
ax.plot([loc[0]], [loc[1]], marker="x", c="black", alpha=0, zorder=1.5)
```

I will make a PR soon.

**Comment by leejjoon** [comment]
A better (and lionger) workaround would be making a rasterization-allowed  AnnotationBbox.

```python
class AnnotationBbox(mpl.offsetbox.AnnotationBbox):
    @mpl.artist.allow_rasterization
    def draw(self, backend):
        super().draw(backend)
```
and then do

```pyton
    abox = AnnotationBbox(
        imbox,
        loc,
    )

```

**Comment by tacaswell** [comment]
Ah, the issue is that the artist in the middle just does not participate in the rasterization logic at all and hence the bug.

Do we need a `forbid_rasterization` decorator that can break the chain?  Thinking through being able to break out of a rasterization stack seems complicated ....

**Comment by leejjoon** [comment]
> Ah, the issue is that the artist in the middle just does not participate in the rasterization logic at all and hence the bug.
> 
> Do we need a `forbid_rasterization` decorator that can break the chain? Thinking through being able to break out of a rasterization stack seems complicated ....

If we want to keep the current behavior that rasterized artists got merged together, I think all artists need to know how to stop rasterization. We may require the draw method of all artists are decorated, or we may use metaclass so that they are decorated by default under the hood.

**Comment by tacaswell** [comment]
I hope `__subclass_init__` is enough to enforce that....

**Comment by QuLogic** [comment]
This was fixed by #24484.
